File: csharp/TetrisGame/Logic/Tetromino/Tetromino.cs
using hu.klenium.tetris.util;
using hu.klenium.tetris.logic.board;
using System;

namespace hu.klenium.tetris.logic.tetromino
{
    public class Tetromino
    {
        private Board board;
        private TetrominoData[] Parts { get; }
        private Point Position { get; set; } = new Point(0, 0);
        private int _rotation = 0;
        private int Rotation {
            get { return _rotation; }
            set { _rotation = (value) % Parts.Length; }
        }
        public Tetromino(int type, Board board)
        {
            this.board = board;
            Parts = TetrominoDataFactory.GetData(type);
        }
        public TetrominoData GetCurrentData()
        {
            return Parts[Rotation];
        }
        public Point GetPosition()
        {
            return Position;
        }

        public bool MoveToInitialPos()
        {
            Dimension boardSize = board.Size;
            float centerOfBoard = boardSize.width / 2.0f;
            Dimension tetrominoBoundingBox = Parts[Rotation].BoundingBox;
            float halfTetrominoWidth = tetrominoBoundingBox.width / 2.0f;
            int centeredTetrominoPosX = (int)Math.Ceiling(centerOfBoard - halfTetrominoWidth);
            return TryPush(new Point(centeredTetrominoPosX, 0));
        }
        public bool RotateRight()
        {
            ++Rotation;
            return true;
        }
        public bool MoveLeft()
        {
            return TryPush(new Point(-1, 0));
        }
        public bool MoveDown()
        {
            return TryPush(new Point(0, 1));
        }
        public bool MoveRight()
        {
            return TryPush(new Point(1, 0));
        }
        
        private bool TryPush(Point delta)
        {
            bool canSlide = CanPushBy(delta);
            if (canSlide)
                Position += delta;
            return canSlide;
        }
        private bool CanPushBy(Point delta)
        {
            return board.CanAddTetromino(this, Position + delta);
        }
    }
}

File: haskell/src/Logic/Tetromino.hs
module Logic.Tetromino (
  Tetromino(..),
  newTetromino,
  moveToInitialPos,
  rotateRight,
  moveLeft,
  moveDown,
  moveRight
) where

import Data.List
import Util.Point
import Util.Dimension
import Util.Boxed
import Logic.Tetromino.Type
import Logic.Tetromino.DataSource
import Logic.Board

newTetromino :: Int -> Tetromino
newTetromino shapeType =
  Tetromino {
    shapeType = shapeType,
    parts = getTetrominoData shapeType 0,
    position = Point 0 0,
    rotation = 0
  }

halfOf :: Int -> Float
halfOf int = (fromIntegral int) / 2

moveToInitialPos :: Tetromino -> Board -> Tetromino
moveToInitialPos tetromino board =
  let centerOfBoard = halfOf $ boundingWidth board
      halfTetrominoWidth = halfOf $ boundingWidth tetromino
      centeredTetrominoPosX = ceiling $ centerOfBoard - halfTetrominoWidth
  in tryPush tetromino board $ Point centeredTetrominoPosX 0

rotateRight :: Tetromino -> Board -> Tetromino
rotateRight original board =
  let nextRotation = (rotation original + 1) `mod` 4
      rotated = original {
        rotation = nextRotation,
        parts = getTetrominoData (shapeType original) nextRotation
      }
  in if canAddTetromino board rotated $ position original
     then rotated
     else let maxPush = boundingWidth rotated
              pushed = multiMoveLeft rotated board 1 maxPush
          in if pushed /= rotated then pushed else original

moveLeft :: Tetromino -> Board -> Tetromino
moveLeft tetromino board =
  tryPush tetromino board (Point (-1) 0)

multiMoveLeft :: Tetromino -> Board -> Int -> Int -> Tetromino
multiMoveLeft tetromino board current max
  | current == max = tetromino
  | otherwise =
      let pushed = tryPush tetromino board (Point (-current) 0)
      in if pushed /= tetromino
         then pushed
         else multiMoveLeft tetromino board (current + 1) max

moveDown :: Tetromino -> Board -> Tetromino
moveDown tetromino board =
  tryPush tetromino board (Point 0 1)

moveRight :: Tetromino -> Board -> Tetromino
moveRight tetromino board =
  tryPush tetromino board (Point 1 0)

tryPush :: Tetromino -> Board -> Point -> Tetromino
tryPush tetromino board delta =
  let newPos = (position tetromino) `plusPoint` delta
  in
    if canAddTetromino board tetromino newPos
    then tetromino { position = newPos }
    else tetromino